RavenDB-13114 : Sharded Hilo

aviv86 opened the pull request:
https://issues.hibernatingrhinos.com/issue/RavenDB-13114

aviv86 committed:
RavenDB-13114 : initial work

aviv86 committed:
RavenDB-13114 : sharded hilo handler

aviv86 committed:
RavenDB-13114 : NextHiLoCommand - avoid passing 'lastMax' query parameter if last range was received from a sharded database -
we don't know if the next range is going to come from the same shard as before and we don't want to "skip" unused numbers

aviv86 committed:
RavenDB-13114 : tests + cleanup

ayende commented on src/Raven.Client/Documents/Commands/HiLoReturnCommand.cs:
```diff
@@ -10,8 +10,9 @@ public class HiLoReturnCommand : RavenCommand
+        private readonly int? _shardIndex;
 
-        public HiLoReturnCommand(string tag, long last, long end)
+        public HiLoReturnCommand(string tag, long last, long end, int? shardIndex = null)
```
Create another ctor, backward compact.

ayende commented on src/Raven.Client/Documents/Commands/HiLoReturnCommand.cs:
```diff
@@ -21,11 +22,12 @@ public HiLoReturnCommand(string tag, long last, long end)
         public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, ServerNode node, out string url)
         {
-            url = $"{node.Url}/databases/{node.Database}/hilo/return?tag={_tag}&end={_end}&last={_last}";
+            url = $"{node.Url}/databases/{node.Database}/hilo/return?tag={_tag}&end={_end}&last={_last}&shardIndex={_shardIndex}";
```
Only if `_shardIndex` is not null

ayende commented on src/Raven.Client/Documents/Commands/NextHiLoCommand.cs:
```diff
@@ -14,27 +16,37 @@ public class NextHiLoCommand : RavenCommand<HiLoResult>
+        private readonly int? _shardIndex;
 
-        public NextHiLoCommand(string tag, long lastBatchSize, DateTime lastRangeAt, char identityPartsSeparator, long lastRangeMax)
+        public NextHiLoCommand(string tag, long lastBatchSize, DateTime lastRangeAt, char identityPartsSeparator, long lastRangeMax, int? shardIndex)
```
different ctor

ayende commented on src/Raven.Client/Documents/Commands/NextHiLoCommand.cs:
```diff
@@ -14,27 +16,37 @@ public class NextHiLoCommand : RavenCommand<HiLoResult>
 
-            var request = new HttpRequestMessage
+            pathBuilder.Append($"{node.Url}/databases/{node.Database}/hilo/next")
+                .Append($"?tag={Uri.EscapeDataString(_tag)}")
```
If you are using StringBuilder, don't user interpolation, write it directly to avoid creating temporaries

ayende commented on src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs:
```diff
@@ -67,15 +73,12 @@ public RangeValue(long min, long max)
         /// <summary>
         /// Generates the document ID.
         /// </summary>
         /// <param name="entity">The entity.</param>
         /// <returns></returns>
-        public async Task<string> GenerateDocumentIdAsync(object entity)
+        public async Task<string> GenerateDocumentIdAsync()
```
You are changing the public API here, why?

aviv86 replied:
`entity` was not used
reverted this

ayende commented on src/Raven.Client/Documents/Identity/AsyncMultiTypeHiLoIdGenerator.cs:
```diff
@@ -48,14 +48,14 @@ public async Task<string> GenerateDocumentIdAsync(object entity)
             var tag = Conventions.TransformTypeCollectionNameToDocumentIdPrefix(typeTagName);
             if (_idGeneratorsByTag.TryGetValue(tag, out var value))
             {
-                return await value.GenerateDocumentIdAsync(entity).ConfigureAwait(false);
```
revert this

ayende commented on src/Raven.Server/Documents/Handlers/HiLoHandler.cs:
```diff
@@ -120,8 +124,10 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
+
                         OldMax = LastRangeMax;
-                        newDoc["Max"] = OldMax + Capacity;
+                        newDoc[Max] = OldMax + Capacity;
```
`nameof` ?

ayende commented on src/Raven.Server/Documents/Handlers/HiLoHandler.cs:
```diff
@@ -98,12 +100,14 @@ internal class MergedNextHiLoCommand : TransactionOperationsMerger.MergedTransac
             public string Prefix;
             public long OldMax;
 
+            private const string Max = "Max";
```
replace with `nameof` ?

ayende commented on src/Raven.Server/Documents/ShardedHandlers/ShardedHiLoHandler.cs:
```diff
@@ -0,0 +1,88 @@
+
+                for (int i = 0; i < numOfShards; i++)
+                {
+                    var cmd = new ShardedCommand(this, Headers.IfMatch);
```
`IfMatch`? we don't do any caching here, no?

ayende commented on src/Raven.Server/Documents/ShardedHandlers/ShardedHiLoHandler.cs:
```diff
@@ -0,0 +1,88 @@
+            using (ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                var tag = GetQueryStringValueAndAssertIfSingleAndNotEmpty("tag");
+                var numOfShards = ShardedContext.RequestExecutors.Length;
```
That isn't number of shards, that is the number of _nodes_.
Note that this _can_ change.

ayende commented on src/Raven.Server/Documents/ShardedHandlers/ShardedHiLoHandler.cs:
```diff
@@ -0,0 +1,88 @@
+                    {
+                        // failed to reach shard
+                        // try the next one
+                        shardIndex = (shardIndex + 1) % numOfShards;
```
I don't think we can rely on this being stable given resharding. We need a stable approach.

ayende commented on src/Raven.Server/Documents/ShardedHandlers/ShardedHiLoHandler.cs:
```diff
@@ -0,0 +1,88 @@
+        [RavenShardedAction("/databases/*/hilo/return", "PUT")]
+        public async Task HiLoReturn()
+        {
+            var shardIndex = GetLongQueryString("shardIndex", required: true);
```
I think a better option would be to use sequential shard ids (0..1M) and use their mapping.
In other words, we'll align that to a shard (and ping between nodes as needed), based to the shard topology.
So the shard index won't be the position here, which may change.

karmeli87 replied:
Another solution might be to move the HiLo doc to be at the system storage of the node.
So it will be node specific only and not shard-specific. All we care here is about non-conflicted ids.
If we do go this way, we need to handle backup/restore properly.

ayende replied:
That would conflict between different sharded databases, and it doesn't match the way we do everything else.
Better to just split it to a shard like we are always doing.

karmeli87 replied:
It doesn't match the way we do everything else indeed :-)
but it will not conflict, since it will be done under write lock of the _server_.

Split to shards might be tricky, since they can move from node and from database.

ayende replied:
That is the _point_, you select a shard id, and return to the shard id.
And the rest is just dispatching to the right node, like we do everywhere here.

aviv86 committed:
RavenDB-13114 : address some PR comments

aviv86 committed:
RavenDB-13114 : implementation changes :
- avoid selecting a random shard to get the hi-low range from, determine the shard to use according to the hilo doc-id
- remove notation of shard-index in client side and in server's response
- generated ids will be the same as in "unsharded" database - without the shard-index suffix

aviv86 committed:
RavenDB-13114 : fix tests

karmeli87 commented on test/FastTests/Client/HiLo.cs:
```diff
@@ -119,7 +118,7 @@ public void Capacity_Should_Double()
 
                     for (var i = 0; i < 32; i++)
-                        hiLoKeyGenerator.GenerateDocumentIdAsync(new User()).GetAwaiter().GetResult();
+                        hiLoKeyGenerator.GenerateDocumentIdAsync(null).GetAwaiter().GetResult();
```
why this change?

aviv86 replied:
reverted this

karmeli87 commented on src/Raven.Client/Documents/Identity/HiLoResult.cs:
```diff
@@ -29,5 +29,12 @@ public class HiLoResult
 
         public DateTime LastRangeAt { get; set; }
 
+        public int? ShardIndex { get; set; }
```
Do we use this?

aviv86 replied:
no we don't
removed

aviv86 committed:
RavenDB-13114 : address PR comments

karmeli87 commented on test/FastTests/Client/HiLo.cs:
```diff
@@ -129,7 +128,7 @@ public void Capacity_Should_Double()
 
                     //we should be receiving a range of 64 now
-                    hiLoKeyGenerator.GenerateDocumentIdAsync(new User()).GetAwaiter().GetResult();
+                    hiLoKeyGenerator.GenerateDocumentIdAsync(null).GetAwaiter().GetResult();
```
revert this as well

ppekrol commented on src/Raven.Client/Documents/Identity/HiLoResult.cs:
```diff
@@ -30,4 +29,9 @@ public class HiLoResult
 
     }
+
+    public class HiloDocument
```
can we make it internal?
